corehq/apps/locations/models.py

```diff
@@ -468,6 +468,39 @@ def get_descendants(self, include_self=False, **kwargs):
             where, **kwargs
         )
 
+    @staticmethod
+    def get_bulk_ancestors(domain, location_ids, **kwargs):
+        """
+        Returns the grouped ancestors for the location ids passed in the
+        dictionary of following pattern
+        {location_id_1: [self, parent, parent_of_parent,.,.,.,],
+        location_id_2: [self, parent, parent_of_parent,.,.,.,],
+
+        }
+        :param domain: domain for which locations is to be pulled out
+        :param location_ids: locations ids whose ancestors needs to be find
+        :param kwargs: extra parameters
+        :return: dict
+        """
+        where = Q(domain=domain, location_id__in=location_ids)
+        location_ancestors = SQLLocation.objects.get_ancestors(where, **kwargs)
+        location_map = {location.location_id: location for location in location_ancestors}
+        location_id_map = {location.id: location for location in location_ancestors}
+
+        grouped_location = {}
+        for location_id in location_ids:
+
+            location_parents = []
+            current_location = location_map[location_id].id if location_id in location_map else None
+
+            while current_location is not None:
+                location_parents.append(location_id_map[current_location])
```

Comment:
can you use the other map here? instead of having two separate ones. the `location_id` exists on each of the location objects in that map

Reply:
No, It looks. Both maps needed here. Because with user object location_id  is available But referencing parent happens using PK. I could do `location.parent.location_id` but that would be way slower to do for all the locations. Therefore I have to refer parent using `location.parent_id` which id a PK therefore I need two dicts